src/anomalib/callbacks/visualizer.py

```diff
@@ -0,0 +1,155 @@
+from anomalib.loggers import AnomalibWandbLogger
+from anomalib.loggers.base import ImageLoggerBase
+from anomalib.models import AnomalyModule
+from anomalib.utils.visualization import (
+    BaseVisualizer,
+    GeneratorResult,
+    VisualizationStep,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class _VisualizationCallback(Callback):
+    def __init__(
+        self,
+        visualizers: BaseVisualizer | list[BaseVisualizer],
+        save: bool = False,
+        save_root: Path | None = None,
+        log: bool = False,
+        show: bool = False,
+    ) -> None:
+        """Callback for visualization that is used internally by the Engine.
+
+        Args:
+            visualizers (BaseVisualizer | list[BaseVisualizer]):
+                Visualizer objects that are used for computing the visualizations. Defaults to None.
+            save (bool, optional): Save the image. Defaults to False.
+            save_root (Path | None, optional): The path to save the images. Defaults to None.
+            log (bool, optional): Log the images into the loggers. Defaults to False.
+            show (bool, optional): Show the images. Defaults to False.
+
+        Example:
+            >>> visualizers = [ImageVisualizer(), MetricsVisualizer()]
+            >>> visualization_callback = _VisualizationCallback(
+            ... visualizers=visualizers,
+            ... save=True,
+            ... save_root="results/images"
+            ... )
+
+            CLI
+            $ anomalib train --model Padim --data MVTec \
+                --visualization.visualizers ImageVisualizer \
+                --visualization.visualizers+=MetricsVisualizer
+
+        Raises:
+            ValueError: Incase `save_root` is None and `save` is True.
+        """
+        self.save = save
+        if save and save_root is None:
+            msg = "`save_root` must be provided if save is True"
+            raise ValueError(msg)
+        self.save_root: Path = save_root if save_root is not None else Path()  # need this check for mypy
+        self.log = log
+        self.show = show
+        self.generators = visualizers if isinstance(visualizers, list) else [visualizers]
+
+    def on_test_batch_end(
+        self,
+        trainer: Trainer,
+        pl_module: AnomalyModule,
+        outputs: STEP_OUTPUT | None,
+        batch: Any,  # noqa: ANN401
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        for generator in self.generators:
+            if generator.visualize_on == VisualizationStep.BATCH:
+                for result in generator(
+                    trainer=trainer,
+                    pl_module=pl_module,
+                    outputs=outputs,
+                    batch=batch,
+                    batch_idx=batch_idx,
+                    dataloader_idx=dataloader_idx,
+                ):
+                    if self.save or self.show:
+                        assert result.file_name is not None, "File name is None"
```

Comment:
is it possible to use an exception handler for this. Bandit keeps complaining about assert